cdap-ui/app/cdap/components/Ingestion/index.tsx

```diff
@@ -0,0 +1,507 @@
+    countAndToggle: {
+      display: 'flex',
+      justifyContent: 'space-between',
+      marginBottom: 10,
+    },
+    toggleView: {
+      color: theme.palette.blue[100],
+      cursor: 'pointer',
+    },
+    configureHeader: {
+      display: 'flex',
+      flexDirection: 'row',
+      justifyContent: 'space-between',
+      alignItems: 'center',
+      padding: 5,
+      paddingLeft: 20,
+      backgroundColor: theme.palette.grey[700],
+    },
+    transferDetailsContainer: {
+      maxWidth: 500,
+      flexDirection: 'column',
+      display: 'flex',
+    },
+    deployTransferBtn: {
+      alignSelf: 'flex-end',
+    },
+  };
+};
+
+interface IIngestionProps extends WithStyles<typeof styles> {
+  test: string;
+  pluginsMap: any;
+}
+
+class IngestionView extends React.Component<IIngestionProps> {
+  public state = {
+    filterStr: '',
+    filteredSources: [],
+    batchsource: [],
+    batchsink: [],
+    selectedSource: null,
+    selectedSink: null,
+    sourceBP: null,
+    sinkBP: null,
+    pipelineName: '',
+    publishingPipeline: false,
+    pipelineDescription: '',
+    modalOpen: false,
+    tableView: true,
+    showConfig: false,
+    pipelineNameError: false,
+    deployFailed: false,
+    cdapVersion: '',
+    currentNs: 'default',
+  };
+
+  public componentDidMount() {
+    const currentNs = getCurrentNamespace();
+    let cdapVersion = VersionStore.getState().version;
+    if (!cdapVersion) {
+      MyCDAPVersionApi.get().subscribe((res) => {
+        cdapVersion = res.version;
+        VersionStore.dispatch({
+          type: VersionActions.updateVersion,
+          payload: {
+            version: res.version,
+          },
+        });
+        this.fetchSourceSink(cdapVersion, currentNs);
+        this.setState({ currentNs, cdapVersion });
+      });
+    } else {
+      this.fetchSourceSink(cdapVersion, currentNs);
+      this.setState({ currentNs, cdapVersion });
+    }
+  }
+
+  public fetchSourceSink = (version, namespace) => {
+    this.fetchPlugins('batchsource', version, namespace);
+    this.fetchPlugins('batchsink', version, namespace);
+  };
+
+  public fetchPlugins = async (extensionType, version, namespace = 'default') => {
+    const res = await MyPipelineApi.fetchPlugins({
+      namespace,
+      pipelineType: 'cdap-data-pipeline',
+      version,
+      extensionType,
+    }).toPromise();
+    const promises = res.map(async (plugin) => {
+      const params = {
+        namespace,
+        artifactName: plugin.artifact.name,
+        artifactVersion: plugin.artifact.version,
+        scope: plugin.artifact.scope,
+        keys: `widgets.${plugin.name}-${plugin.type}`,
+      };
+      const widgetJ = await MyPipelineApi.fetchWidgetJson(params).toPromise();
+      const widgetJson = JSON.parse(widgetJ[`widgets.${plugin.name}-${plugin.type}`]);
+      const pluginWithWidgetJson = { ...plugin, widgetJson };
+      return pluginWithWidgetJson;
+    });
+
+    Promise.all(promises).then((data) => {
```

Comment:
all the APIs are already observable. You can use combineLatest instead of converting everything to promises

Reply:
Added a TODO, will cleanup in the upcoming days (post 6.2.0).